File: backend/spire-api-dotnet-aspire/SpireCore.API/JWT/Identity/JwtIdentityService.cs
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;

namespace SpireCore.API.JWT.Identity;

public class JwtIdentityService : IJwtIdentityService
{
    private readonly IConfiguration _config;
    private readonly byte[] _key;

    public JwtIdentityService(IConfiguration config)
    {
        _config = config;
        _key = Encoding.UTF8.GetBytes(_config["Jwt:Key"]!);
    }

    public string GenerateJwt(
        IJwtIdentity identity,
        IEnumerable<Claim>? extraClaims = null,
        int? expiresInMinutes = null)
    {
        // Base claims
        var claims = new List<Claim>
        {
            new Claim(JwtRegisteredClaimNames.Sub, identity.Id.ToString()),
            new Claim(JwtRegisteredClaimNames.Iss, _config["Jwt:Issuer"]!),
            new Claim(JwtRegisteredClaimNames.Aud, _config["Jwt:Audience"]!),
            new Claim(JwtRegisteredClaimNames.Iat,
                      DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(),
                      ClaimValueTypes.Integer64)
        };

        // User vs Service
        if (!identity.IsService && identity is IJwtUserIdentity u)
        {
            if (!string.IsNullOrEmpty(u.Email))
                claims.Add(new Claim("email", u.Email));
            if (!string.IsNullOrEmpty(u.UserName))
                claims.Add(new Claim("userName", u.UserName));
            if (!string.IsNullOrEmpty(u.DisplayName))
                claims.Add(new Claim("displayName", u.DisplayName));
            if (!string.IsNullOrEmpty(u.FirstName))
                claims.Add(new Claim("firstName", u.FirstName));
            if (!string.IsNullOrEmpty(u.LastName))
                claims.Add(new Claim("lastName", u.LastName));
            if (!string.IsNullOrEmpty(u.ImageUrl))
                claims.Add(new Claim("image_url", u.ImageUrl));
        }
        else if (identity.IsService && identity is IJwtServiceIdentity s)
        {
            claims.Add(new Claim("client_id", s.ServiceName));
            foreach (var scope in s.Scopes)
                claims.Add(new Claim("scope", scope));
            if (!string.IsNullOrEmpty(s.ImageUrl))
                claims.Add(new Claim("image_url", s.ImageUrl));
        }

        // Extras
        if (extraClaims != null)
            claims.AddRange(extraClaims);

        // Lifetime
        var defaultMin = int.TryParse(_config["Jwt:AccessTokenMinutes"], out var cfgMin)
                            ? cfgMin : 1440;
        var exp = DateTime.UtcNow.AddMinutes(expiresInMinutes ?? defaultMin);

        // Create token
        var creds = new SigningCredentials(
                        new SymmetricSecurityKey(_key),
                        SecurityAlgorithms.HmacSha256);

        var token = new JwtSecurityToken(
            claims: claims,
            notBefore: DateTime.UtcNow,
            expires: exp,
            signingCredentials: creds);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public ClaimsPrincipal? ValidateJwt(string token)
    {
        token = token?.Trim().Trim('"');

        var handler = new JwtSecurityTokenHandler();
        var parms = new TokenValidationParameters
        {
            ValidateIssuer = true,
            ValidateAudience = true,
            ValidateIssuerSigningKey = true,
            ValidateLifetime = true,
            ValidIssuer = _config["Jwt:Issuer"],
            ValidAudience = _config["Jwt:Audience"],
            IssuerSigningKey = new SymmetricSecurityKey(_key),
            ClockSkew = TimeSpan.FromMinutes(2)
        };

        try
        {
            return handler.ValidateToken(token, parms, out _);
        }
        catch
        {
            return null;
        }
    }

    public IJwtIdentity? GetIdentityFromClaims(ClaimsPrincipal principal)
    {
        if (principal?.Identity?.IsAuthenticated != true)
            return null;

        var sub = principal.FindFirstValue(JwtRegisteredClaimNames.Sub)
              ?? principal.FindFirstValue(ClaimTypes.NameIdentifier);

        if (!Guid.TryParse(sub, out var id))
            return null;

        var raw = ToUniqueClaimDict(principal.Claims);

        var imageUrl = principal.FindFirstValue("image_url"); // Or whatever claim you use

        // service tokens carry client_id
        var clientId = principal.FindFirstValue("client_id");
        if (!string.IsNullOrEmpty(clientId))
            return new JwtServiceIdentity
            {
                Id = id,
                Issuer = principal.FindFirstValue(JwtRegisteredClaimNames.Iss)
                         ?? principal.FindFirstValue("iss")!,
                RawClaims = raw,
                ServiceName = clientId,
                Scopes = principal.FindAll("scope").Select(c => c.Value).ToList(),
                ImageUrl = imageUrl
            };

        // otherwise user token
        return new JwtUserIdentity
        {
            Id = id,
            Issuer = principal.FindFirstValue(JwtRegisteredClaimNames.Iss)
                     ?? principal.FindFirstValue("iss")!,
            RawClaims = raw,
            Email = principal.FindFirstValue("email")
                         ?? principal.FindFirstValue(ClaimTypes.Email),
            UserName = principal.FindFirstValue("userName")
                         ?? principal.FindFirstValue(ClaimTypes.Name),
            DisplayName = principal.FindFirstValue("displayName"),
            FirstName = principal.FindFirstValue("firstName"),
            LastName = principal.FindFirstValue("lastName"),
            ImageUrl = imageUrl
        };
    }

    public IJwtUserIdentity? GetUserFromClaims(ClaimsPrincipal principal)
        => GetIdentityFromClaims(principal) as IJwtUserIdentity;

    public IJwtServiceIdentity? GetServiceFromClaims(ClaimsPrincipal principal)
        => GetIdentityFromClaims(principal) as IJwtServiceIdentity;

    public Guid? GetIdFromToken(string jwtToken)
    {
        var principal = ValidateJwt(jwtToken);
        return principal is null
            ? null
            : Guid.Parse(principal.FindFirstValue(JwtRegisteredClaimNames.Sub));
    }

    public bool IsTokenValid(string jwtToken)
        => ValidateJwt(jwtToken) is not null;

    public bool IsExpired(string jwtToken)
    {
        var handler = new JwtSecurityTokenHandler();
        if (!handler.CanReadToken(jwtToken))
            return true;

        var token = handler.ReadJwtToken(jwtToken);
        return token.ValidTo < DateTime.UtcNow;
    }

    private static IReadOnlyDictionary<string, object> ToUniqueClaimDict(
    IEnumerable<Claim> claims)
    {
        var dict = new Dictionary<string, object>(StringComparer.Ordinal);
        foreach (var c in claims)
        {
            if (!dict.ContainsKey(c.Type))
                dict[c.Type] = c.Value;
        }
        return dict;
    }

}

File: backend/spire-api-nestjs/src/spire-core/jwt/jwt.identity.service.ts
// src/core/jwt/jwt.identity.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtPayload, sign, verify, VerifyOptions } from 'jsonwebtoken';
import {
  IJwtIdentity, IJwtServiceIdentity, IJwtUserIdentity,
  JwtServiceIdentity, JwtUserIdentity,
} from './jwt.identity';

type Claims = Record<string, unknown>;

@Injectable()
export class JwtIdentityService {
  private readonly key: string;
  private readonly issuer: string;
  private readonly audience: string;
  private readonly defaultMinutes: number;

  constructor(private readonly config: ConfigService) {
    this.key = this.config.get<string>('JWT_KEY', '')!;
    this.issuer = this.config.get<string>('JWT_ISSUER', '')!;
    this.audience = this.config.get<string>('JWT_AUDIENCE', '')!;
    this.defaultMinutes = Number(this.config.get<string>('JWT_ACCESS_MINUTES', '1440'));
    if (!this.key || !this.issuer || !this.audience) {
      throw new Error('JWT_KEY / JWT_ISSUER / JWT_AUDIENCE must be configured');
    }
  }

  generateJwt(identity: IJwtIdentity, extraClaims?: Claims, expiresInMinutes?: number): string {
    const iatSec = Math.floor(Date.now() / 1000);
    const expSec = iatSec + 60 * (expiresInMinutes ?? this.defaultMinutes);

    const base: Claims = { sub: identity.id, iss: this.issuer, aud: this.audience, iat: iatSec, exp: expSec };

    if (!identity.isService) {
      const u = identity as IJwtUserIdentity;
      if (u.email) base.email = u.email;
      if (u.userName) base.userName = u.userName;
      if (u.displayName) base.displayName = u.displayName;
      if (u.firstName) base.firstName = u.firstName;
      if (u.lastName) base.lastName = u.lastName;
      if (u.imageUrl) base.image_url = u.imageUrl;
    } else {
      const s = identity as IJwtServiceIdentity;
      base.client_id = s.serviceName;
      if (s.scopes?.length) base.scope = s.scopes;
      if (s.imageUrl) base.image_url = s.imageUrl;
    }

    const payload: Claims = { ...base, ...(extraClaims ?? {}) };
    return sign(payload, this.key, { algorithm: 'HS256' });
  }

  validateJwt(token?: string): JwtPayload | null {
    if (!token) return null;
    const cleaned = token.trim().replace(/^Bearer\s+/i, '').replace(/^"|"$/g, '');
    const opts: VerifyOptions = {
      algorithms: ['HS256'],
      audience: this.audience,
      issuer: this.issuer,
      clockTolerance: 120, // 2 minutes
    };
    try {
      const decoded = verify(cleaned, this.key, opts);
      return typeof decoded === 'object' ? (decoded as JwtPayload) : null;
    } catch {
      return null;
    }
  }

  getIdentityFromPayload(payload: JwtPayload | null): IJwtIdentity | null {
    if (!payload || !payload.sub) return null;
    const raw = this.toUniqueClaimDict(payload);

    const issuer = (payload.iss as string) ?? '';
    const id = String(payload.sub);
    const imageUrl = (payload['image_url'] as string | undefined) ?? null;
    const clientId = (payload['client_id'] as string | undefined) ?? '';

    if (clientId) {
      const scopesRaw = payload['scope'];
      const scopes = Array.isArray(scopesRaw)
        ? scopesRaw.map(String)
        : typeof scopesRaw === 'string'
          ? [scopesRaw]
          : [];
      const svc = new JwtServiceIdentity();
      svc.id = id;
      svc.issuer = issuer;
      svc.rawClaims = raw;
      svc.serviceName = clientId;
      svc.scopes = scopes;
      svc.imageUrl = imageUrl;
      return svc;
    }

    const usr = new JwtUserIdentity();
    usr.id = id;
    usr.issuer = issuer;
    usr.rawClaims = raw;
    usr.email = (payload['email'] as string | undefined) ?? null;
    usr.userName = (payload['userName'] as string | undefined) ?? null;
    usr.displayName = (payload['displayName'] as string | undefined) ?? null;
    usr.firstName = (payload['firstName'] as string | undefined) ?? null;
    usr.lastName = (payload['lastName'] as string | undefined) ?? null;
    usr.imageUrl = imageUrl;
    return usr;
  }

  getUserFromPayload(payload: JwtPayload | null): IJwtUserIdentity | null {
    const id = this.getIdentityFromPayload(payload);
    return id?.isService ? null : (id as IJwtUserIdentity);
  }

  getServiceFromPayload(payload: JwtPayload | null): IJwtServiceIdentity | null {
    const id = this.getIdentityFromPayload(payload);
    return id?.isService ? (id as IJwtServiceIdentity) : null;
  }

  getIdFromToken(token: string): string | null {
    const p = this.validateJwt(token);
    return p?.sub ? String(p.sub) : null;
  }

  isTokenValid(token: string): boolean {
    return this.validateJwt(token) !== null;
  }

  isExpired(token: string): boolean {
    try {
      const payload = this.decode(token);
      if (!payload?.exp) return true;
      return payload.exp * 1000 < Date.now();
    } catch {
      return true;
    }
  }

  getClaim(token: string, claimType: string): string | undefined {
    const p = this.decode(token);
    const val = p?.[claimType];
    return typeof val === 'string' ? val : undefined;
  }

  private decode(token: string): JwtPayload | null {
    const cleaned = token.trim().replace(/^Bearer\s+/i, '').replace(/^"|"$/g, '');
    try {
      const base64 = cleaned.split('.')[1];
      if (!base64) return null;
      const json = Buffer.from(base64, 'base64').toString('utf8');
      return JSON.parse(json);
    } catch {
      return null;
    }
  }

  private toUniqueClaimDict(obj: Record<string, unknown>): Readonly<Record<string, unknown>> {
    const dict: Record<string, unknown> = {};
    Object.keys(obj).forEach(k => {
      if (!(k in dict)) dict[k] = obj[k] as unknown;
    });
    return dict;
  }
}
